File: java/cucumber-archetype/src/test/java/postoffice/stepdefs/CalculateShippingCostSteps.java
package postoffice.stepdefs;

import io.cucumber.java.en.Given;
import io.cucumber.java.en.When;
import io.cucumber.java.en.Then;
import static org.junit.Assert.*;

import postoffice.models.Package;
import postoffice.models.DeliveryType;
import postoffice.services.LabelService;

public class CalculateShippingCostSteps {

    private Package pkg;
    private LabelService labelService = new LabelService();
    private double calculatedShippingCost;

    @Given("a package with dimensions for shipping {int} height, {int} width, {int} length, and weight {double} kg, and delivery type {string}")
    public void calculateShippingCostPackage(int height, int width, int length, double weight, String deliveryType) {
        pkg = new Package(); // Initialize a new Package object

        pkg.setHeight(height);
        pkg.setWidth(width);
        pkg.setLength(length);
        pkg.setWeight(weight);
        pkg.setDeliveryType(DeliveryType.valueOf(deliveryType.toUpperCase())); // Convert string to DeliveryType enum
    }

    @When("the shipping cost is calculated for the package")
    public void whenTheShippingCostIsCalculated() {
        calculatedShippingCost = labelService.calculateShippingCost(pkg);
    }

    @Then("the shipping cost should be {double}")
    public void thenTheShippingCostShouldBe(double expectedCost) {
        assertEquals("Shipping cost should match the expected cost", expectedCost, calculatedShippingCost, 0.01);
    }
}

File: node/PostOffice/src/features/support/CalculateShippingCostSteps.ts
import {Given, Then, When} from "@cucumber/cucumber";
import {Package} from "../../models/Package";
import {DeliveryType} from "../../models/DeliveryType";
import {LabelService} from "../../services/LabelService";
import assert from "assert";

let pkg: Package;
let actualCost: number;

Given('a package with dimensions for shipping {int} height, {int} width, {int} length, and weight {float} kg, and delivery type {string}', function (height, width, length, weight, deliveryType) {
  const deliveryEnum = DeliveryType[deliveryType as keyof typeof DeliveryType];
  if (!deliveryEnum) {
    throw new Error(`Invalid delivery type: ${deliveryType}`);
  }
  pkg = new Package("Id", height, width, length, weight, deliveryEnum);
});

When('the shipping cost is calculated for the package', function () {
  actualCost = LabelService.calculateShippingCost(pkg);
});

Then('the shipping cost should be {float}', function (expectedCost) {
  assert.equal(expectedCost, actualCost, `Expected cost: ${expectedCost}, Actual cost: ${actualCost}`);
});

